internal/controlplane/xds_clusters.go

```diff
@@ -1,70 +1,138 @@
 	"github.com/golang/protobuf/ptypes"
 
 	"github.com/pomerium/pomerium/config"
+	"github.com/pomerium/pomerium/internal/log"
 	"github.com/pomerium/pomerium/internal/urlutil"
 )
 
 func (srv *Server) buildClusters(options *config.Options) []*envoy_config_cluster_v3.Cluster {
 	grpcURL := &url.URL{
-		Scheme: "grpc",
+		Scheme: "http",
 		Host:   srv.GRPCListener.Addr().String(),
 	}
 	httpURL := &url.URL{
 		Scheme: "http",
 		Host:   srv.HTTPListener.Addr().String(),
 	}
 	authzURL := &url.URL{
-		Scheme: strings.Replace(options.AuthorizeURL.Scheme, "http", "grpc", -1),
+		Scheme: options.AuthorizeURL.Scheme,
 		Host:   options.AuthorizeURL.Host,
 	}
 
 	clusters := []*envoy_config_cluster_v3.Cluster{
-		srv.buildCluster("pomerium-control-plane-grpc", grpcURL),
-		srv.buildCluster("pomerium-control-plane-http", httpURL),
-		srv.buildCluster("pomerium-authz", authzURL),
+		srv.buildInternalCluster("pomerium-control-plane-grpc", grpcURL, true),
+		srv.buildInternalCluster("pomerium-control-plane-http", httpURL, false),
+		srv.buildInternalCluster("pomerium-authz", authzURL, true),
 	}
 
 	if config.IsProxy(options.Services) {
-		type clusterDestination struct {
-			name, scheme, hostport string
-		}
-		clusterDestinations := map[clusterDestination]struct{}{}
 		for _, policy := range options.Policies {
-			name, scheme, hostport := srv.getClusterDetails(policy.Destination)
-			clusterDestinations[clusterDestination{name, scheme, hostport}] = struct{}{}
+			clusters = append(clusters, srv.buildPolicyCluster(&policy))
 		}
+	}
+
+	return clusters
+}
 
-		for dst := range clusterDestinations {
-			name, scheme, hostport := dst.name, dst.scheme, dst.hostport
-			clusters = append(clusters, srv.buildCluster(name, &url.URL{
-				Scheme: scheme,
-				Host:   hostport,
-			}))
+func (srv *Server) buildInternalCluster(name string, endpoint *url.URL, forceHTTP2 bool) *envoy_config_cluster_v3.Cluster {
+	var transportSocket *envoy_config_core_v3.TransportSocket
+	if endpoint.Scheme == "https" {
+		transportSocket = &envoy_config_core_v3.TransportSocket{
+			Name: "tls",
 		}
 	}
+	return srv.buildCluster(name, endpoint, transportSocket, forceHTTP2)
+}
+
+func (srv *Server) buildPolicyCluster(policy *config.Policy) *envoy_config_cluster_v3.Cluster {
+	name := getPolicyName(policy)
+	return srv.buildCluster(name, policy.Destination, srv.buildPolicyTransportSocket(policy), false)
+}
+
+func (srv *Server) buildPolicyTransportSocket(policy *config.Policy) *envoy_config_core_v3.TransportSocket {
+	if policy.Destination.Scheme != "https" {
+		return nil
+	}
 
-	return clusters
+	sni := policy.Destination.Hostname()
+	if policy.TLSServerName != "" {
+		sni = policy.TLSServerName
+	}
+	tlsContext := &envoy_extensions_transport_sockets_tls_v3.UpstreamTlsContext{
+		CommonTlsContext: &envoy_extensions_transport_sockets_tls_v3.CommonTlsContext{
+			AlpnProtocols: []string{"http/1.1"},
```

Comment:
This is where we need an additional policy setting to explicitly enable http2 right?

Reply:
Yes and this too:

```
cluster.Http2ProtocolOptions = &envoy_config_core_v3.Http2ProtocolOptions{}
```